File: rust/bundlebase/benches/bench_functions/double_val.go
// Benchmark IPC function server for Go. Implements double_val (scalar) and int_sum (aggregate).
package main

import (
	"os"

	sdk "github.com/bundlebase/bundlebase/sdk/go/bundlebasesdk"
	"github.com/apache/arrow-go/v18/arrow"
	"github.com/apache/arrow-go/v18/arrow/array"
	"github.com/apache/arrow-go/v18/arrow/memory"
)

type doubleVal struct{}

func (d *doubleVal) Invoke(batch arrow.Record) (arrow.Array, error) {
	input := batch.Column(0).(*array.Int64)
	alloc := memory.NewGoAllocator()
	b := array.NewInt64Builder(alloc)
	for i := 0; i < input.Len(); i++ {
		b.Append(input.Value(i) * 2)
	}
	return b.NewArray(), nil
}

type intSum struct{}

func (s *intSum) CreateState() (interface{}, error) {
	return int64(0), nil
}

func (s *intSum) Accumulate(state interface{}, batch arrow.Record) (interface{}, error) {
	sum := state.(int64)
	input := batch.Column(0).(*array.Int64)
	for i := 0; i < input.Len(); i++ {
		sum += input.Value(i)
	}
	return sum, nil
}

func (s *intSum) Merge(stateA interface{}, stateB interface{}) (interface{}, error) {
	return stateA.(int64) + stateB.(int64), nil
}

func (s *intSum) Evaluate(state interface{}) (interface{}, error) {
	return state.(int64), nil
}

type benchProvider struct{}

func (p *benchProvider) Functions() map[string]sdk.FunctionRef {
	return map[string]sdk.FunctionRef{
		"double_val": {Scalar: &doubleVal{}},
		"int_sum":    {Aggregate: &intSum{}},
	}
}

func (p *benchProvider) Metadata() sdk.FunctionManifest {
	return sdk.FunctionManifest{
		Functions: []sdk.FunctionMeta{
			{Name: "double_val", InputTypes: []string{"Int64"}, ReturnType: "Int64", Kind: "scalar"},
			{Name: "int_sum", InputTypes: []string{"Int64"}, ReturnType: "Int64", Kind: "aggregate"},
		},
	}
}

func main() {
	_ = os.Stderr // Suppress unused import
	sdk.ServeFunction(&benchProvider{})
}

File: rust/bundlebase/benches/bench_functions/DoubleVal.java
// Benchmark IPC function server for Java. Implements double_val (scalar) and int_sum (aggregate).

import com.bundlebase.sdk.Function;
import com.bundlebase.sdk.FunctionServe;
import org.apache.arrow.memory.RootAllocator;
import org.apache.arrow.vector.BigIntVector;
import org.apache.arrow.vector.FieldVector;
import org.apache.arrow.vector.VectorSchemaRoot;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class DoubleVal {

    static class DoubleValFn implements Function.ScalarFunction {
        @Override
        public FieldVector invoke(VectorSchemaRoot input) {
            BigIntVector col = (BigIntVector) input.getVector(0);
            BigIntVector result = new BigIntVector("result", new RootAllocator());
            result.allocateNew(col.getValueCount());
            for (int i = 0; i < col.getValueCount(); i++) {
                result.setSafe(i, col.get(i) * 2);
            }
            result.setValueCount(col.getValueCount());
            return result;
        }
    }

    static class IntSumFn implements Function.AggregateFunction<Long> {
        @Override
        public Long createState() {
            return 0L;
        }

        @Override
        public Long accumulate(Long state, VectorSchemaRoot input) {
            BigIntVector col = (BigIntVector) input.getVector(0);
            long sum = state;
            for (int i = 0; i < col.getValueCount(); i++) {
                sum += col.get(i);
            }
            return sum;
        }

        @Override
        public Long merge(Long stateA, Long stateB) {
            return stateA + stateB;
        }

        @Override
        public Object evaluate(Long state) {
            return state;
        }
    }

    static class BenchProvider implements Function.FunctionProvider {
        @Override
        public Map<String, Object> functions() {
            Map<String, Object> fns = new LinkedHashMap<>();
            fns.put("double_val", new DoubleValFn());
            fns.put("int_sum", new IntSumFn());
            return fns;
        }

        @Override
        public Function.FunctionManifest metadata() {
            return new Function.FunctionManifest(List.of(
                new Function.FunctionMeta("double_val", List.of("Int64"), "Int64", "scalar"),
                new Function.FunctionMeta("int_sum", List.of("Int64"), "Int64", "aggregate")
            ));
        }
    }

    public static void main(String[] args) {
        FunctionServe.run(new BenchProvider(), args);
    }
}
